### PR TITLE
Feat/controller refactor

### PR DESCRIPTION
**Pull request by angelxmoreno**
## Summary by Sourcery

Refactor the API controllers into a generic BaseController with lifecycle hooks and standardized responses, update WorkspacesController to enforce permissions through hooks, simplify routes, switch to DSN-based database configuration, adjust database services and entities, and update tests and documentation to match the new patterns.

New Features:
- Introduce a generic BaseController with built-in CRUD methods, before/after hooks, zod validation, and standardized single and paginated response types

Enhancements:
- Refactor WorkspacesController to override BaseController hooks for permission checks and post-creation permission seeding
- Streamline API routing to only mount workspace endpoints under a dedicated route
- Switch database configuration to use a single DATABASE_URL DSN and update shared-core config and pagination types accordingly
- Update BaseDbService and DbServiceInterface to merge and return updated entities

Build:
- Adjust apps/api package.json with hot reload dev script and add sqlite3 and typescript-config dependencies

Documentation:
- Update database and shared-core README files to document DSN format, query params, and configuration examples

Tests:
- Refactor integration test helpers for new response formats, update API tests to assert expected status codes, add a database connection test in init suite
- Add migrations for user_permissions schema changes and SQLite compatibility

Chores:
- Add unique constraint on UserPermissionEntity(userId, workspaceId) and remove legacy composite key
- Remove unused controllers and exports, and update test fixtures with additional entity fields

### CODE REVIEW

File: apps/api/src/controllers/WorkspacesController.ts [hunk 1]
```diff
@@ -1,101 +1,83 @@
...
+    protected override async afterCreate({
+        context,
+        entity,
+    }: ResultHookParams<WorkspaceEntity>): Promise<SuccessResponseSingle<WorkspaceEntity>> {
+        const user = this.getCurrentUser(context);
+        await this.userPermissionsDbService.save({
+            userId: user.id,
+            workspaceId: entity.id,
+            canWrite: true,
+        });
```
[the remark this answers is not in the document]
Comment (angelxmoreno) [reply apps/api/src/controllers/WorkspacesController.ts:80]: I disagree with this suggestion. While error handling is important, the suggested implementation is overly complex for our current needs. The proposed solution adds significant complexity for handling race conditions that are not a pressing concern at this stage. Simple error handling would be sufficient if we determine it's needed in the future, but this level of defensive programming is premature optimization.

File: apps/api/tests/integration/auth.test.ts [hunk 2]
```diff
@@ -47,20 +46,17 @@ describe('API Authentication', () => {
                 headers: { Authorization: 'InvalidFormat dev_admin_key_0123456789abcdef0123456789abcdef01234567' },
             });
```
[the remark this answers is not in the document]
Comment (angelxmoreno) [reply apps/api/tests/integration/auth.test.ts:49]: I disagree with this suggestion. The test is currently checking the expected behavior (status code 400), and adding error message validation is premature optimization. The test should focus on the functional behavior rather than implementation-specific error messages. If we want to standardize error message formats, that should be done as a separate initiative with consistent error message standards across the entire API.

File: apps/api/tests/integration/auth.test.ts [hunk 3]
```diff
@@ -47,20 +46,17 @@ describe('API Authentication', () => {
...
 
-            expect(response.status).toBe(400);
             await expectErrorResponse(response, 400);
         });
 
         it('should work with different user types', async () => {
             // Test with john user
```
[the remark this answers is not in the document]
Comment (angelxmoreno) [reply apps/api/tests/integration/auth.test.ts:55]: Excellent suggestion! Testing edge cases for users with no permissions is valuable for ensuring robust permission handling. I've added a test case for the 'jack' user who has no workspace permissions to verify that our permission system correctly handles this scenario.
